File: apps/server/app/domain/playback_state.py
"""Domain types for shared playback state.

Pure Python: no framework, no I/O, no wall clock (timestamps are plain data
set by adapters). Moved from app/services/stream_engine.py during the
clean-architecture migration; app.services.stream_engine re-exports these for
compatibility.
"""

from __future__ import annotations

from dataclasses import dataclass
from enum import StrEnum


class PlaybackMode(StrEnum):
    idle = "idle"
    playing = "playing"


class RepeatMode(StrEnum):
    off = "off"
    all = "all"
    one = "one"


@dataclass
class PlaybackState:
    mode: PlaybackMode = PlaybackMode.idle
    now_playing_id: int | None = None
    now_playing_title: str | None = None
    now_playing_channel: str | None = None
    now_playing_thumbnail_url: str | None = None
    now_playing_duration_seconds: int | None = None
    now_playing_is_live: bool = False
    started_at_epoch_seconds: float | None = None
    started_at_monotonic_seconds: float | None = None
    paused: bool = False
    paused_elapsed_seconds: float | None = None
    repeat_mode: RepeatMode = RepeatMode.off
    shuffle_enabled: bool = False

File: packages/domain/src/playback-state.ts
/**
 * Domain layer (pure TypeScript, zero dependencies).
 *
 * Ported from the Python app/domain package (clean-architecture migration);
 * same rules apply: no I/O, no wall clock, no framework imports. Time and
 * randomness arrive as arguments. Shared by the Node server and the web
 * client — this is the single source of truth for playback rules.
 */

export const PLAYBACK_MODE_IDLE = "idle" as const;
export const PLAYBACK_MODE_PLAYING = "playing" as const;
export type PlaybackModeValue = typeof PLAYBACK_MODE_IDLE | typeof PLAYBACK_MODE_PLAYING;

export const REPEAT_OFF = "off" as const;
export const REPEAT_ALL = "all" as const;
export const REPEAT_ONE = "one" as const;
export type RepeatModeValue = typeof REPEAT_OFF | typeof REPEAT_ALL | typeof REPEAT_ONE;

export interface PlaybackState {
  mode: PlaybackModeValue;
  nowPlayingId: number | null;
  nowPlayingTitle: string | null;
  nowPlayingChannel: string | null;
  nowPlayingThumbnailUrl: string | null;
  nowPlayingDurationSeconds: number | null;
  nowPlayingIsLive: boolean;
  startedAtEpochSeconds: number | null;
  startedAtMonotonicSeconds: number | null;
  paused: boolean;
  /** True from the moment a track is chosen until its audio pipeline spawns. */
  loading: boolean;
  pausedElapsedSeconds: number | null;
  repeatMode: RepeatModeValue;
  shuffleEnabled: boolean;
}

export function initialPlaybackState(): PlaybackState {
  return {
    mode: PLAYBACK_MODE_IDLE,
    nowPlayingId: null,
    nowPlayingTitle: null,
    nowPlayingChannel: null,
    nowPlayingThumbnailUrl: null,
    nowPlayingDurationSeconds: null,
    nowPlayingIsLive: false,
    startedAtEpochSeconds: null,
    startedAtMonotonicSeconds: null,
    paused: false,
    loading: false,
    pausedElapsedSeconds: null,
    repeatMode: REPEAT_OFF,
    shuffleEnabled: false,
  };
}
